lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemStats.java

```diff
@@ -0,0 +1,264 @@
+    {
+        return listObjectsV2;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getPutObject()
+    {
+        return putObject;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getDeleteObject()
+    {
+        return deleteObject;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getDeleteObjects()
+    {
+        return deleteObjects;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getCreateMultipartUpload()
+    {
+        return createMultipartUpload;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getCompleteMultipartUpload()
+    {
+        return completeMultipartUpload;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getAbortMultipartUpload()
+    {
+        return abortMultipartUpload;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats uploadPart()
+    {
+        return uploadPart;
+    }
+
+    public MetricPublisher newMetricPublisher()
+    {
+        return new JmxMetricPublisher(this);
+    }
+
+    public static final class JmxMetricPublisher
+            implements MetricPublisher
+    {
+        private static final Set<SdkMetric<?>> ALLOWED_METRICS = Set.of(API_CALL_SUCCESSFUL, RETRY_COUNT, API_CALL_DURATION, ERROR_TYPE);
+
+        private static final Logger log = Logger.get(JmxMetricPublisher.class);
+
+        private final S3FileSystemStats stats;
+
+        public JmxMetricPublisher(S3FileSystemStats stats)
+        {
+            this.stats = requireNonNull(stats, "stats is null");
+        }
+
+        @Override
+        public void publish(MetricCollection metricCollection)
+        {
+            try {
+                Optional<String> serviceId = metricCollection.metricValues(SERVICE_ID).stream().filter(Objects::nonNull).findFirst();
+                Optional<String> operationName = metricCollection.metricValues(OPERATION_NAME).stream().filter(Objects::nonNull).findFirst();
+                if (serviceId.isEmpty() || operationName.isEmpty()) {
+                    log.warn("ServiceId or OperationName is empty for AWS MetricCollection: %s", metricCollection);
+                    return;
+                }
+
+                if (!serviceId.get().equals("S3")) {
+                    return;
+                }
+
+                AwsSdkV2ApiCallStats apiCallStats = getApiCallStats(operationName.get());
+                publishMetrics(metricCollection, apiCallStats);
+            }
+            catch (Exception e) {
+                log.warn(e, "Publishing AWS metrics failed");
+            }
+        }
+
+        private void publishMetrics(MetricCollection metricCollection, AwsSdkV2ApiCallStats apiCallStats)
+        {
+            metricCollection.stream()
+                    .filter(metricRecord -> metricRecord.metric() != null && metricRecord.value() != null)
+                    .filter(metricRecord -> ALLOWED_METRICS.contains(metricRecord.metric()))
```

Comment:
Minor: `ALLOWED_METRICS.contains(metricRecord.metric())` will eliminate `null `values, making the check above slightly redundant. Since this is in a relatively hot path (per SDK call), we should prefer combining these two filters into a single one like: `metricRecord -> metricRecord.value() != null && ALLOWED_METRICS.contains(metricRecord.metric())`

Reply:
Thanks! Updated in the revision